unittests/finality_tests.hpp

```diff
@@ -0,0 +1,188 @@
+         vote.strong = true;
+      } else {
+         vote.strong = false;
+      }
+      return node1.control->process_vote_message( vote );
+   }
+
+   // send node2's vote identified by "index" in the collected votes
+   eosio::chain::vote_status process_node2_vote(uint32_t index, vote_mode mode = vote_mode::strong) {
+      FC_ASSERT( index < node2_votes.size(), "out of bound index in process_node2_vote" );
+      return process_vote( node2_votes[index], mode );
+   }
+
+   // send node2's latest vote
+   eosio::chain::vote_status process_node2_vote(vote_mode mode = vote_mode::strong) {
+      auto index = node2_votes.size() - 1;
+      return process_vote( node2_votes[index], mode );
+   }
+
+   // send node3's vote identified by "index" in the collected votes
+   eosio::chain::vote_status process_node3_vote(uint32_t index, vote_mode mode = vote_mode::strong) {
+      FC_ASSERT( index < node3_votes.size(), "out of bound index in process_node3_vote" );
+      return process_vote( node3_votes[index], mode );
+   }
+
+   // send node3's latest vote
+   eosio::chain::vote_status process_node3_vote(vote_mode mode = vote_mode::strong) {
+      auto index = node3_votes.size() - 1;
+      return process_vote( node3_votes[index], mode );
+   }
+
+   // returns true if node1's LIB has advanced
+   bool node1_lib_advancing() {
+      return lib_advancing(node1.control->if_irreversible_block_num(), node1_prev_lib_bum);
+   }
+
+   // returns true if node2's LIB has advanced
+   bool node2_lib_advancing() {
+      return lib_advancing(node2.control->if_irreversible_block_num(), node2_prev_lib_bum);
+   }
+
+   // returns true if node3's LIB has advanced
+   bool node3_lib_advancing() {
+      return lib_advancing(node3.control->if_irreversible_block_num(), node3_prev_lib_bum);
+   }
+
+   // Produces a number of blocks and returns true if LIB is advancing.
+   // This function can be only used at the end of a test as it clears
+   // node2_votes and node3_votes when starting.
+   bool produce_blocks_and_verify_lib_advancing() {
+      // start from fresh
+      node2_votes.clear();
+      node3_votes.clear();
+
+      for (auto i = 0; i < 3; ++i) {
+         produce_and_push_block();
+         process_node2_vote();
+         if (!node1_lib_advancing() || !node2_lib_advancing() || !node3_lib_advancing()) {
+            return false;
+         }
+      }
+
+      return true;
+   }
+
+   std::vector<eosio::chain::vote_message> node2_votes;
+
+private:
+   bool lib_advancing(uint32_t curr_lib_num, uint32_t& prev_lib_num) {
+      auto advancing = curr_lib_num > prev_lib_num;
+      // update pre_lib_num for next time check
+      prev_lib_num = curr_lib_num;
+      return advancing;
+   }
+
+   eosio::testing::tester   node1;
+   eosio::testing::tester   node2;
+   eosio::testing::tester   node3;
+   uint32_t                 node1_prev_lib_bum {0};
+   uint32_t                 node2_prev_lib_bum {0};
+   uint32_t                 node3_prev_lib_bum {0};
```

Comment:
Having a `std::array<node_info, 3>` with 
```
struct node_info {
    eosio::testing::tester   node;
    uint32_t                 node_prev_lib_bum;
};
```

would remove code duplication (in member functions and elsewhere).

Reply:
Thanks. Done

Reply:
Also move the implementation out of .hpp file.

Reply:
Somewhat minor, but I wonder if you shouldn't name them `node0, node1, node2` if you use the array approach so they match index.

Reply:
Yes, I will make the change. That reduces confusion about which node is which. Thanks.